File: app/src/main/java/com/cn29/aac/ui/masterdetail/SimpleListActivity.java
package com.cn29.aac.ui.masterdetail;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import com.cn29.aac.R;
import com.cn29.aac.ui.base.BaseAppCompatActivity;
import com.cn29.aac.ui.masterdetail.vm.SimpleMasterDetailShareViewModel;
import javax.inject.Inject;


/**
 * An activity representing a list of FeedEntrys. This activity
 * has different presentations for handset and tablet-size devices. On
 * handsets, the activity presents a list of items, which when touched,
 * lead to a {@link SimpleDetailActivity} representing
 * item details. On tablets, the activity presents the list of items and
 * item details side-by-side using two vertical panes.
 * https://developer.android.com/topic/libraries/architecture/viewmodel.html
 */
public class SimpleListActivity extends BaseAppCompatActivity {

  @Inject
  SimpleMasterDetailShareViewModel masterDetailShareViewModel;

  /**
   * Whether or not the activity is in two-pane mode, i.e. running on a tablet
   * device.
   */
  private boolean mTwoPane;

  private GithubRepoAdapter githubRepoAdapter;

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_simple_list);
    Toolbar toolbar = findViewById(R.id.toolbar);
    setSupportActionBar(toolbar);
    toolbar.setTitle(getTitle());
    FloatingActionButton fab = findViewById(R.id.fab);
    fab.setOnClickListener(view -> loadRepos("googlesamples"));
    setupAdapter();
    loadRepos("googlesamples");
    RecyclerView recyclerView = findViewById(R.id.feedentry_list);
    assert recyclerView != null;
    recyclerView.setAdapter(githubRepoAdapter);
    if (findViewById(R.id.feedentry_detail_container) != null) {
      // The detail container view will be present only in the
      // large-screen layouts (res/values-w900dp).
      // If this view is present, then the
      // activity should be in two-pane mode.
      mTwoPane = true;
    }
  }

  private void loadRepos(String ownerName) {
    masterDetailShareViewModel
        .loadRepos(ownerName)
        .observe(SimpleListActivity.this, repos -> {
          assert repos != null;
          switch (repos.status) {
            case SUCCESS:
              progressDialogComponent.hideLoading();
              githubRepoAdapter.setRepos(repos.data);
              break;
            case ERROR:
              progressDialogComponent.hideLoading();
              break;
            case LOADING:
              progressDialogComponent.showLoading();
              break;
          }
          githubRepoAdapter.notifyDataSetChanged();
        });
  }

  private void setupAdapter() {
    githubRepoAdapter = new GithubRepoAdapter(
        repo -> {
          if (mTwoPane) {
            Bundle arguments = new Bundle();
            arguments.putString(SimpleDetailFragment.REPO_NAME, repo.name);
            arguments.putString(SimpleDetailFragment.OWNER_NAME, repo.owner.login);
            SimpleDetailFragment fragment = new SimpleDetailFragment();
            fragment.setArguments(arguments);
            getSupportFragmentManager().beginTransaction()
                .replace(R.id.feedentry_detail_container, fragment)
                .commit();
          } else {
            Context context = SimpleListActivity.this;
            Intent intent = new Intent(context, SimpleDetailActivity.class);
            intent.putExtra(SimpleDetailFragment.REPO_NAME, repo.name);
            intent.putExtra(SimpleDetailFragment.OWNER_NAME, repo.owner.login);
            context.startActivity(intent);
          }
        });
  }
}

File: app/src/main/java/com/cn29/aac/ui/masterdetail/SimpleListActivity.kt
package com.cn29.aac.ui.masterdetail

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.Toast
import androidx.appcompat.widget.Toolbar
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.RecyclerView
import com.cn29.aac.R
import com.cn29.aac.repo.github.Repo
import com.cn29.aac.ui.base.BaseAppCompatActivity
import com.cn29.aac.ui.common.BaseRecyclerViewAdapter.OnItemClickListener
import com.cn29.aac.ui.masterdetail.vm.SimpleMasterDetailShareViewModel
import com.cn29.aac.util.Result
import com.google.android.material.floatingactionbutton.FloatingActionButton
import javax.inject.Inject

class SimpleListActivity : BaseAppCompatActivity() {
    @Inject
    lateinit var masterDetailShareViewModel: SimpleMasterDetailShareViewModel

    private var mTwoPane = false
    private lateinit var githubRepoAdapter: GithubRepoAdapter
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_simple_list)
        val toolbar = findViewById<Toolbar>(
                R.id.toolbar)
        setSupportActionBar(toolbar)
        toolbar.title = title
        val fab = findViewById<FloatingActionButton>(R.id.fab)
        fab.setOnClickListener {
            loadRepos("googlesamples")
        }
        setupAdapter()
        loadRepos("googlesamples")
        val recyclerView = findViewById<RecyclerView>(R.id.feedentry_list)!!
        recyclerView.adapter = githubRepoAdapter
        if (findViewById<View?>(R.id.feedentry_detail_container) != null) {
            // The detail container view will be present only in the
            // large-screen layouts (res/values-w900dp).
            // If this view is present, then the
            // activity should be in two-pane mode.
            mTwoPane = true
        }
    }

    private fun loadRepos(ownerName: String) {
        masterDetailShareViewModel
                .loadRepos(ownerName)
                .observe(this, Observer {
                    when (it) {
                        Result.Loading -> {
                            progressDialogComponent?.showLoading()
                            Toast.makeText(this, "Loading", Toast.LENGTH_SHORT)
                                    .show()
                        }
                        is Result.Success -> {
                            progressDialogComponent?.hideLoading()
                            githubRepoAdapter.setRepos(it.data)
                            Toast.makeText(this, "Success", Toast.LENGTH_SHORT)
                                    .show()
                        }
                        is Result.Error -> progressDialogComponent?.hideLoading()
                        else -> progressDialogComponent?.hideLoading()
                    }
                    githubRepoAdapter.notifyDataSetChanged()
                })
    }

    private fun setupAdapter() {
        val listener = object : OnItemClickListener<Repo?> {
            override fun onItemClick(item: Repo?) {

                if (mTwoPane) {
                    val arguments = Bundle()
                    arguments.putString(
                            SimpleDetailFragment.REPO_NAME,
                            item!!.name)
                    arguments.putString(
                            SimpleDetailFragment.OWNER_NAME,
                            item.owner.login)
                    val fragment = SimpleDetailFragment()
                    fragment.arguments = arguments
                    supportFragmentManager.beginTransaction()
                            .replace(
                                    R.id.feedentry_detail_container,
                                    fragment)
                            .commit()
                } else {
                    val context: Context = this@SimpleListActivity
                    val intent = Intent(
                            context,
                            SimpleDetailActivity::class.java)
                    intent.putExtra(
                            SimpleDetailFragment.REPO_NAME,
                            item!!.name)
                    intent.putExtra(
                            SimpleDetailFragment.OWNER_NAME,
                            item.owner.login)
                    context.startActivity(
                            intent)
                }

            }
        }
        githubRepoAdapter = GithubRepoAdapter(listener)
    }
}